enhancements/oc/olm-mirroring.md

```diff
@@ -0,0 +1,115 @@
+# OC Tooling for Disconnected OLM
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+This enhancement outlines the changes necessary to `oc` to support disconnected distribution of OLM catalogs. It relies on the [Disconnected OLM enhancement](https://github.com/operator-framework/enhancements/blob/master/enhancements/opm-mirroring.md).
+
+## Motivation
+
+Disconnected installation of OLM catalogs can be done with the following `opm` commands:
+
+```
+$ opm registry build build --appregistry-namespace=quay.io/community-operators --to disconnected-registry:5000/openshift/community-operators-catalog:4.2.1 
+pulling quay.io/community-operators/etcd...
+buliding catalog...
+mirroring catalog...
+quay.io/community-operators mirrored to disconnected-registry:5000/openshift/community-operators-catalog:4.2.1
+
+$ docker pull disconnected-registry:5000/openshift/community-operators-catalog:4.2.1
+
+$ opm registry images --from=disconnected-registry:5000/openshift/community-operators-catalog:4.2.1 --to=disconnected-registry:5000/community-operators --manifests=./mirror-manifests | xargs oc image mirror
+
+$ oc apply -f ./mirror-manifests
+```
+
+This requires both `oc` and `opm` to be installed and available on bastion hosts. We would like to distribute one tool for creating disconnected tools, not two.
+
+### Goals
+* Define `oc` commands which can build and mirror all images required for mirroring to a disconnected envorinment
+
+### Non-Goals
+* Provide an api for fine-grained control over which operators / operands are mirrored. 
+
+## Proposal
+
+The basic approach will be to:
+
+- Generate catalog container images that can replace the appregistry repositories
+- For each operator within the catalog image, mirror it to the disconnected environment
+
+### Generate and mirror a catalog
+
+`oc` will be extended with a command (equivalent to the `opm` command)
+
+```sh
+$ oc catalog build --appregistry-namespace=quay.io/community-operators --to quay.io/ecordell/community-operators-catalog:4.2.1 
```

Comment:
This seems reasonable, might be better for `oc adm` as I comment below.

The use of `-namespace` here is a bit confusing.  Could this instead be done with a wildcard?  Is there a use case where I might want to do:

```
oc adm catalog build --appregistry quay.io/community-operators/foo --appregistry quay.io/community-operators/bar ...
```

instead?  If so, then the current flag might be better as:

```
oc adm catalog build --appregistry quay.io/community-operators/* ...
```

Reply:
I think that we had this as a non-goal:

> * Provide an api for fine-grained control over which operators / operands are mirrored. 

If we are ever going to support this, then we may want to make the command as described.

Reply:
It's actually a slightly more complex problem to make this easy to configure -- right now the `https://quay.io/cnr/` endpoint is the entrypoint into that entire API, and everything else is defined with a relatively complex set of query parameters in order to drill down into individual operators. That means that the only easily human readable way to read those requests is to do something like:

```
$ oc catalog build --appregistry-namespace=community-operators --operators=foo,bar,baz
```

we could explicitly expect a wildcard:
```
$ oc catalog build --appregistry-namespace=community-operators --operators=* 
```

or we could assume that not setting that field implies "everything in the namespace"
```
$ oc catalog build --appregistry-namespace=community-operators
```

If we want to give an option to use something other than the `cnr` endpoint to interact with the registry, we would also need to have a separate parameter for that:

```
$ oc catalog build --appregistry-namespace=community-operators --endpoint="https://otherdomain.io/differentendpoint"
```

Not the prettiest thing, but since the client does some fairly complex abstraction around writing query parameters and (I believe) stuff in the body of certain requests doing it any other way would probably require us taking a look at the implementation of the app-registry API itself.

Reply:
All of those options require an understanding of dependencies that I think is unreasonable for users. (`--operators=foo,bar` - does `foo` depend on `baz`? How do you know?) 

I think we should keep it to namespaces for now; we can build much better ux around the catalog images once we're off of appregistry.

> If we want to give an option to use something other than the cnr endpoint to interact with the registry, we would also need to have a separate parameter for that:

There is an `--appregistry-endpoint` flag in this proposal for that case.

Reply:
as long as appregistry is clearly denoted, the concepts make sense, and you have no plans for further extension (because appregistry is dead and the flags would be deprecated), this is fine as long as we make it clear that namespace is not kube related.

Reply:
`appregistry-organization`?

Reply:
Changed to `appregistry-org`, with the description clarifying.